### PR TITLE
Convert np scalars

### LINKED ISSUE #124: MathArrays behave badly with numpy's numeric types
Behold, my nightmare:

```python
from mitxgraders.helpers.calc import evaluator
variables = {'v': MathArray([3, 4])}
evaluator('v*v + v') # scalar + vector,
                     # but does not throw error
```

## Explanation:

```python
import numpy as np
from mitxgraders.helpers.calc import MathArray

x = MathArray([1, 2, 3])

1.0 + x                 # throws an error, good!
np.float64(1.0) + x     # returns MathArray([2.0, 3.0, 4.0])
```

Here's what's happening:

- `1.0 + x`:
    - tries to call `(1.0).__add__(self, x)` but fails, because builtin float type does not know how to handle `MathArray`
    - next tries `x.__radd__(self, 1.0)`, which knows that it should through a StudentFacingError.
- `np.float64(1.0) + x`:
    - tries to call `np.float64(1.0).__add__(self, x)` and succeeds: Numpy's `float64` numeric type knows how to add `ndarray`s to itself.
    - I.e., this version never calls any of `MathArray`'s methods

I'm guessing this is true for all numeric operators and true for all numpy numeric types. Bummer.

Note: `np.float64.__add__` is implemented in C. So even if we wanted to monkey-patch it, I don't think we could.

## Goals:

  - Resolve this issue (insofar as it could ever affect students) 
  - continue to keep most of the type/shape checking outside of calc.py

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
Resolves #124 

Also implements #123 without errors.

### CODE REVIEW

File: mitxgraders/helpers/calc/calc.py [hunk 1]
```diff
@@ -591,7 +620,7 @@ def handle_node(node):
             if any(np.any(np.isnan(r)) for r in as_list):
                 return float('nan')
 
-            return result
+            return cast_np_numeric_as_builtin(result)
```
Comment (jolyonb) [inline mitxgraders/helpers/calc/calc.py]: Shouldn't this be mapped across lists?

**Comment by jolyonb** [comment]
Looks good. Just the one comment, I think.

File: mitxgraders/helpers/calc/calc.py (hunk 1 above)
Comment (ChristopherChudzicki) [reply mitxgraders/helpers/calc/calc.py]: I agree—good idea—and it's done. That said, I couldn't find a place where it mattered. (`result` should only be a list if it is the result of `eval_arguments`, and in that case all the list items should already have passed through `handle_node`). Anyway, I agree, but that's why there's no new test asserting this behavior.

File: mitxgraders/helpers/calc/calc.py [hunk 2]
```diff
@@ -591,7 +630,7 @@ def handle_node(node):
             if any(np.any(np.isnan(r)) for r in as_list):
                 return float('nan')
 
-            return result
+            return cast_np_numeric_as_builtin(result, map_across_lists=False)
```
Comment (jolyonb) [inline mitxgraders/helpers/calc/calc.py]: Did you mean to set this True?

File: mitxgraders/helpers/calc/calc.py (hunk 1 above)
Comment (jolyonb) [reply mitxgraders/helpers/calc/calc.py]: Agreed, I'm not sure it matters presently, but future-proofing!

File: mitxgraders/helpers/calc/calc.py (hunk 2 above)
Comment (ChristopherChudzicki) [reply mitxgraders/helpers/calc/calc.py]: Oops, that was a relic from trying to find a test that would break when `False`. Fixed.

**Pull request merged**